### PR TITLE
feat: Remove batch locks from non-atomic squashing

### PR DESCRIPTION
**Pull request by dranikpg**
This PR changes the locking mechanism of non atomic squashing

Squashed sequences consists of batches that are executed in parallel on multiple shards.

Previously, all keys of a batch were collected and locked ahead of executing the full batch on multiple shards (akin to running a transaction with those collected keys)

Now, no keys are collected in advance. Instead the commands are just sent to their respective shards where they're executed on separate transactions. We could create a new transaction for every run, but I just create a single non-atomic multi transaction for every shard.

This new locking mechanism will be used for parallelizing pipelines.

### CODE REVIEW

File: src/server/multi_command_squasher.cc [hunk 1]
```diff
@@ -26,18 +26,24 @@ template <typename F> void IterateKeys(CmdArgList args, KeyIndex keys, F&& f) {
...
-  if (!sinfo.local_tx)
-    sinfo.local_tx = new Transaction{cntx_->transaction};
+  if (!sinfo.local_tx) {
+    if (base_cid_) {
+      sinfo.local_tx = new Transaction{cntx_->transaction};
+    } else {
+      sinfo.local_tx = new Transaction{cntx_->transaction->GetCId(), sid};
+      sinfo.local_tx->StartMultiNonAtomic();
+    }
+  }
```
Comment (dranikpg) [inline src/server/multi_command_squasher.cc:46]: I feel like this needs a comment, by I've already commented twice on the difference of non-atomic/atomic

File: src/server/multi_command_squasher.cc [hunk 2]
```diff
@@ -202,6 +207,15 @@ void MultiCommandSquasher::Run() {
...
         [this](ShardId sid) { return sharded_[sid].had_writes; });
+
+  // UnlockMulti is a no-op for non-atomic multi transactions,
+  // still called for correctness and future changes
+  if (base_cid_ == nullptr) {
+    for (auto& sd : sharded_) {
+      if (sd.local_tx)
+        sd.local_tx->UnlockMulti();
+    }
+  }
```
Comment (dranikpg) [inline src/server/multi_command_squasher.cc:218]: Questionable but I added it

File: src/server/multi_command_squasher.cc (hunk 1 above)
Comment (romange) [reply src/server/multi_command_squasher.cc:46]: you can add a comment where you comment about it :)

**Comment by romange** [comment]
How this change affect the usual Dragonfly run with default flags? I assume we do not ask for non-atomic transactions explicitly.

**Comment by dranikpg** [comment]
I doesn't affect "the default" at all. Non atomic multi runs only if
1. A script runs non-atomically
2. Multi exec mode is non atomic

I guess the previous version should have been slightly faster, but more problematic due to heavier locks. I think  its not that critical as the main user of non-atomic multi will now be pipelines where we decided not to combine locks

File: src/server/multi_command_squasher.cc [hunk 3]
```diff
@@ -138,22 +141,25 @@ bool MultiCommandSquasher::ExecuteSquashed() {
...
     return false;
 
+  for (auto& sd : sharded_)
+    sd.replies.reserve(sd.cmds.size());
+
   Transaction* tx = cntx_->transaction;
 
-  if (track_keys_) {
-    tmp_keylist_.assign(collected_keys_.begin(), collected_keys_.end());
-    tx->PrepareSquashedMultiHop(base_cid_, CmdArgList{tmp_keylist_});
```
Comment (chakaz) [inline src/server/multi_command_squasher.cc:145]: Do you want to delete the overload of `PrepareSquashedMultiHop()` that accepts keys?

File: src/server/multi_command_squasher.cc [hunk 4]
```diff
@@ -26,18 +26,24 @@ template <typename F> void IterateKeys(CmdArgList args, KeyIndex keys, F&& f) {
...
 
 MultiCommandSquasher::ShardExecInfo& MultiCommandSquasher::PrepareShardInfo(ShardId sid) {
   if (sharded_.empty())
     sharded_.resize(shard_set->size());
 
   auto& sinfo = sharded_[sid];
-  if (!sinfo.local_tx)
-    sinfo.local_tx = new Transaction{cntx_->transaction};
+  if (!sinfo.local_tx) {
+    if (base_cid_) {
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/multi_command_squasher.cc]: Done

File: src/server/multi_command_squasher.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/multi_command_squasher.cc:145]: Done

File: src/server/multi_command_squasher.cc [hunk 5]
```diff
@@ -202,6 +208,19 @@ void MultiCommandSquasher::Run() {
   if (!sharded_.empty())
     cntx_->transaction->ReportWritesSquashedMulti(
         [this](ShardId sid) { return sharded_[sid].had_writes; });
+
+  // UnlockMulti is a no-op for non-atomic multi transactions,
+  // still called for correctness and future changes
+  if (base_cid_ == nullptr) {
```
Comment (chakaz) [inline src/server/multi_command_squasher.cc:214]: Use `!IsAtomic()` instead?

Comment (dranikpg) [reply src/server/multi_command_squasher.cc]: Forgot that one 😨

**Review by chakaz — APPROVED** [review]
LGTM, but it might be worth getting another approval from @romange to make sure I didn't miss anything :)

**Review by romange — APPROVED** [review]

**Pull request merged**